From acc4a3b91faa37ac0b90be4ede639baed58786d0 Mon Sep 17 00:00:00 2001
From: Xiaoliang Yang <xiaoliang.yang_1@nxp.com>
Date: Wed, 24 Aug 2022 10:18:23 +0800
Subject: [PATCH 1/2] lldpd: add additional ethernet capabilitie tlv

Additional Ethernet capabilities TLV are defined in IEEE Std 802.3-2018.
Its subtype is 7, and the first 4 bits indicate preemption capability
and status. The preemption status is got from ethtool ioctl in Linux.

Signed-off-by: Xiaoliang Yang <xiaoliang.yang_1@nxp.com>
---
 include/linux/ethtool.h       | 13 +++++++++++++
 src/daemon/interfaces-linux.c | 31 +++++++++++++++++++++++++++++++
 src/daemon/lldp-tlv.h         |  1 +
 src/daemon/protocols/lldp.c   | 13 +++++++++++++
 src/lldpd-structs.h           |  9 +++++++++
 5 files changed, 67 insertions(+)

diff --git a/include/linux/ethtool.h b/include/linux/ethtool.h
index 6bfbb85..65201e2 100644
--- a/include/linux/ethtool.h
+++ b/include/linux/ethtool.h
@@ -341,6 +341,16 @@ struct ethtool_eee {
 	__u32	reserved[2];
 };
 
+struct ethtool_fp {
+	__u32   cmd;
+	__u8    fp_supported;
+	__u8    fp_enabled;
+	__u32   supported_queues_mask;
+	__u32   preemptible_queues_mask;
+	__u32   min_frag_size;
+	__u32   reserved[2];
+};
+
 /**
  * struct ethtool_modinfo - plugin module eeprom information
  * @cmd: %ETHTOOL_GMODULEINFO
@@ -1396,6 +1406,9 @@ enum ethtool_fec_config_bits {
 #define ETHTOOL_GFECPARAM	0x00000050 /* Get FEC settings */
 #define ETHTOOL_SFECPARAM	0x00000051 /* Set FEC settings */
 
+#define ETHTOOL_GFP             0x00000052 /* Get Frame Preemption settings */
+#define ETHTOOL_SFP             0x00000053 /* Set Frame Preemption settings */
+
 /* compatibility with older code */
 #define SPARC_ETH_GSET		ETHTOOL_GSET
 #define SPARC_ETH_SSET		ETHTOOL_SSET
diff --git a/src/daemon/interfaces-linux.c b/src/daemon/interfaces-linux.c
index 5f901e6..ef7f3ab 100644
--- a/src/daemon/interfaces-linux.c
+++ b/src/daemon/interfaces-linux.c
@@ -42,6 +42,8 @@
 #define MAX_PORTS 1024
 #define MAX_BRIDGES 1024
 
+#define DIV_ROUND_UP(n, d) (((n)+(d)-1) / (d))
+
 static int
 iflinux_eth_init(struct lldpd *cfg, struct lldpd_hardware *hardware)
 {
@@ -545,6 +547,32 @@ end:
 	return rc;
 }
 
+static int
+iflinux_ethtool_gfp(struct lldpd *cfg, const char *ifname, struct lldpd_dot3_fp *p_fp) {
+	int rc;
+	struct ethtool_fp gfp;
+	struct ifreq ifr = {};
+
+	strlcpy(ifr.ifr_name, ifname, sizeof(ifr.ifr_name));
+
+	memset(&gfp, 0, sizeof(gfp));
+	gfp.cmd = ETHTOOL_GFP;
+	ifr.ifr_data = (caddr_t)&gfp;
+	rc = ioctl(cfg->g_sock, SIOCETHTOOL, &ifr);
+	if (rc) {
+		log_warnx("interfaces", "get preemption error");
+
+		return rc;
+	}
+
+	p_fp->support = gfp.fp_supported;
+	p_fp->status = gfp.fp_enabled;
+	p_fp->active = gfp.fp_enabled;
+	p_fp->frag_size = DIV_ROUND_UP(gfp.min_frag_size, 64) - 1;
+
+	return 0;
+}
+
 /* Fill up MAC/PHY for a given hardware port */
 static void
 iflinux_macphy(struct lldpd *cfg, struct lldpd_hardware *hardware)
@@ -566,6 +594,9 @@ iflinux_macphy(struct lldpd *cfg, struct lldpd_hardware *hardware)
 
 	log_debug("interfaces", "ask ethtool for the appropriate MAC/PHY for %s",
 	    hardware->h_ifname);
+
+	iflinux_ethtool_gfp(cfg, hardware->h_ifname, &port->p_fp);
+
 	if (iflinux_ethtool_glink(cfg, hardware->h_ifname, &uset) == 0) {
 		port->p_macphy.autoneg_support = iflinux_ethtool_link_mode_test_bit(
 			ETHTOOL_LINK_MODE_Autoneg_BIT, uset.link_modes.supported);
diff --git a/src/daemon/lldp-tlv.h b/src/daemon/lldp-tlv.h
index a8016bb..3bafcb8 100644
--- a/src/daemon/lldp-tlv.h
+++ b/src/daemon/lldp-tlv.h
@@ -46,6 +46,7 @@
 #define LLDP_TLV_DOT3_POWER	2
 #define LLDP_TLV_DOT3_LA	3
 #define LLDP_TLV_DOT3_MFS	4
+#define LLDP_TLV_DOT3_FP       7
 
 #define LLDP_TLV_MED_CAP	1
 #define LLDP_TLV_MED_POLICY	2
diff --git a/src/daemon/protocols/lldp.c b/src/daemon/protocols/lldp.c
index a74556a..49b1ee6 100644
--- a/src/daemon/protocols/lldp.c
+++ b/src/daemon/protocols/lldp.c
@@ -81,6 +81,7 @@ static int _lldp_send(struct lldpd *global,
 #endif
 #ifdef ENABLE_DOT3
 	const u_int8_t dot3[] = LLDP_TLV_ORG_DOT3;
+	u_int8_t fp_tlv;
 #endif
 #ifdef ENABLE_LLDPMED
 	int i;
@@ -283,6 +284,18 @@ static int _lldp_send(struct lldpd *global,
 	      POKE_END_LLDP_TLV))
 		goto toobig;
 
+	fp_tlv = port->p_fp.support | port->p_fp.status << 1 |
+		 port->p_fp.active << 2 | port->p_fp.frag_size << 3;
+	if (!(
+	      POKE_START_LLDP_TLV(LLDP_TLV_ORG) &&
+	      POKE_BYTES(dot3, sizeof(dot3)) &&
+	      POKE_UINT8(LLDP_TLV_DOT3_FP) &&
+	      POKE_UINT16(fp_tlv) &&
+	      POKE_END_LLDP_TLV))
+		goto toobig;
+
+
+
 	/* MAC/PHY */
 	if (!(
 	      POKE_START_LLDP_TLV(LLDP_TLV_ORG) &&
diff --git a/src/lldpd-structs.h b/src/lldpd-structs.h
index 88d347c..f9b73cb 100644
--- a/src/lldpd-structs.h
+++ b/src/lldpd-structs.h
@@ -142,6 +142,14 @@ struct lldpd_dot3_power {
 	u_int8_t		pd_load;
 	u_int16_t		pse_max;
 };
+
+struct lldpd_dot3_fp {
+	u_int8_t		support;
+	u_int8_t		status;
+	u_int8_t		active;
+	u_int8_t		frag_size;
+};
+
 MARSHAL(lldpd_dot3_power);
 #endif
 
@@ -282,6 +290,7 @@ struct lldpd_port {
 	u_int32_t		 p_aggregid;
 	struct lldpd_dot3_macphy p_macphy;
 	struct lldpd_dot3_power	 p_power;
+	struct lldpd_dot3_fp     p_fp;
 #endif
 
 #ifdef ENABLE_LLDPMED
-- 
2.17.1

